### PR TITLE
[Diagnostics] Improve {{none}} fix-it verifier

### PR DESCRIPTION
**Pull request by omochi**
This PR improve {{none}} fix-it verifier.

Currently, it means that no fix-it is expected.

I improve it to mean that no more fix-it is expected when use it with normal fix-it verifier at same time.

For example:

```swift
func labeledFunc(aa: Int, bb: Int) {}

func test() {
  // (1)
  labeledFunc(aax: 0, bb: 1) // expected-error {{incorrect argument label in call (have 'aax:bb:', expected 'aa:bb:')}} {{15-18=aa}}
  
  // (2)
  labeledFunc(aax: 0, bbx: 1) // expected-error {{incorrect argument labels in call (have 'aax:bbx:', expected 'aa:bb:')}} {{15-18=aa}}
}
```

They means that diagnosed fix-its include `{{15-18=aa}}` at least.
They doesn't care about whether more other fix-its are emitted.
Actually (2) emits second fix-its for label `bbx:`, but it doesn't affect anything.

I introduce new feature for such situation to check strictly fix-its are just equal to expected.
For example:

```swift
func test() {
  // (3)
  labeledFunc(aax: 0, bb: 1) // expected-error {{incorrect argument label in call (have 'aax:bb:', expected 'aa:bb:')}} {{15-18=aa}} {{none}}
  
  // (4)
  labeledFunc(aax: 0, bbx: 1) // expected-error {{incorrect argument labels in call (have 'aax:bbx:', expected 'aa:bb:')}} {{15-18=aa}} {{none}}
}
```

With above code, (3) is passed verification but (4) is not.

We can check fix-its more strictly.

---

Original code doesn't support using it with normal fix-its.

This is result from Xcode11.4

```swift
a.swift:3:135: error: expected no fix-its; actual fix-it seen: {{13-16=aa}} {{21-24=bb}}
labeledFunc(aax: 0, bbx: 1) // expected-error {{incorrect argument labels in call (have 'aax:bbx:', expected 'aa:bb:')}} {{13-16=aa}} {{none}}
                                                                                                                                      ^~~~~~~
                                                                                                                                      {{13-16=aa}} {{21-24=bb}}
```

---

@CodaFi @gregomni @owenv Could you review this?

### CODE REVIEW

**Comment by omochi** [comment]
@swift-ci Please smoke test

**Review by owenv — COMMENTED** [review]
I don't know how often this will be used, but I think you're right that it could be useful for some of the diagnostics like argument renaming which produce many fix-its. I have a few comments which I think could help improve the clarity of the code.

File: test/diagnostics/verifier.swift
Comment (LucianoPAlmeida) [inline test/diagnostics/verifier.swift:19]: Does this make sense to be `actual fix-its seen:` because it can have more than one? :)

**Comment by omochi** [comment]
I have exactly been working on improving the diagnosis of argument labels recently.
I wanted this feature because I had to fix a lot of test cases in that work.

File: lib/Frontend/DiagnosticVerifier.cpp
Comment (omochi) [reply lib/Frontend/DiagnosticVerifier.cpp:571]: OK 🙂
Thanks for example.

File: lib/Frontend/DiagnosticVerifier.cpp
Comment (omochi) [reply lib/Frontend/DiagnosticVerifier.cpp:567]: The reason I have merged the two cases is that both conditions hold in the following case. 
https://github.com/apple/swift/blob/422e7012209c4117019c162bad508e463bdd6b46/test/Frontend/verify-fixits.swift#L71

Thank you for pointing out that it's hard to understand.
I will try a split.

File: lib/Frontend/DiagnosticVerifier.cpp
Comment (omochi) [reply lib/Frontend/DiagnosticVerifier.cpp:611]: OK

File: lib/Frontend/DiagnosticVerifier.cpp
Comment (omochi) [reply lib/Frontend/DiagnosticVerifier.cpp:612]: This part corresponds to the following test cases.

https://github.com/apple/swift/blob/422e7012209c4117019c162bad508e463bdd6b46/test/Frontend/verify-fixits.swift#L11
https://github.com/apple/swift/blob/422e7012209c4117019c162bad508e463bdd6b46/test/Frontend/verify-fixits.swift#L19

Fix-its removes unwanted ` {{1-1=a}}`.

How do you think?

File: lib/Frontend/DiagnosticVerifier.cpp
Comment (omochi) [reply lib/Frontend/DiagnosticVerifier.cpp:606]: You are right. Thank you.

File: test/diagnostics/verifier.swift
Comment (omochi) [reply test/diagnostics/verifier.swift:19]: Thanks. I will support plural form.

**Comment by omochi** [comment]
@owenv @LucianoPAlmeida 

I updated implementation.

I followed the points except for the following.
https://github.com/apple/swift/pull/30791#discussion_r403243835

I think it's getting better.

File: lib/Frontend/DiagnosticVerifier.cpp
Comment (owenv) [inline lib/Frontend/DiagnosticVerifier.cpp:568]: Nit: I think this would be cleaner if it returned an `std::pair` of `actualFixitsStr` and the phrase instead of using `actualFixitsStr` as an out parameter.

File: lib/Frontend/DiagnosticVerifier.cpp
Comment (owenv) [inline lib/Frontend/DiagnosticVerifier.cpp:608]: Nit: a comment explaining why this is needed might be nice for future readers.

File: lib/Frontend/DiagnosticVerifier.cpp
Comment (owenv) [inline lib/Frontend/DiagnosticVerifier.cpp:576]: This is fine, but using `llvm::Twine` might make the concatenation a little more efficient

File: lib/Frontend/DiagnosticVerifier.cpp
Comment (owenv) [inline lib/Frontend/DiagnosticVerifier.cpp:624]: I don't think the verifier currently requires `{{none}}` is at the end. I think it probably makes sense to report an error if it does not. I think we can diagnose this just before we start parsing a fix-it on line 469 if Expected.noExtraFixitsMayAppear is already true.

**Review by owenv — COMMENTED** [review]
I like the new direction, I just have a few more small comments. Also, thanks for adding all the new tests!

File: lib/Frontend/DiagnosticVerifier.cpp
Comment (owenv) [reply lib/Frontend/DiagnosticVerifier.cpp]: I see, thanks for explaining this

**Comment by omochi** [comment]
Thank you for review. I will check tomorrow.

File: lib/Frontend/DiagnosticVerifier.cpp
Comment (omochi) [reply lib/Frontend/DiagnosticVerifier.cpp]: OK. I did it.

File: lib/Frontend/DiagnosticVerifier.cpp
Comment (omochi) [reply lib/Frontend/DiagnosticVerifier.cpp:631]: I added comment.
https://github.com/apple/swift/blob/efaacddecf0b309215ff8e41b24d73558c64d6b8/lib/Frontend/DiagnosticVerifier.cpp#L604-L615

I also added here.
https://github.com/apple/swift/blob/efaacddecf0b309215ff8e41b24d73558c64d6b8/lib/Frontend/DiagnosticVerifier.cpp#L650-L661

File: lib/Frontend/DiagnosticVerifier.cpp
Comment (omochi) [reply lib/Frontend/DiagnosticVerifier.cpp]: Thanks! I did it.

File: lib/Frontend/DiagnosticVerifier.cpp
Comment (omochi) [reply lib/Frontend/DiagnosticVerifier.cpp]: I thought that check is needed too.
I was going to make it another PR.

I implemented it and added test case.

**Comment by omochi** [comment]
I updated 😄

@swift-ci Please smoke test

File: lib/Frontend/DiagnosticVerifier.cpp
Comment (owenv) [inline lib/Frontend/DiagnosticVerifier.cpp:469]:
```suggestion
                       "}} was found. It may only appear once in an expectation.");
```

**Review by owenv — APPROVED** [review]
I left one optional comment for rewording a message, but this looks great. Thank you!

**Comment by omochi** [comment]
@owenv Thanks for your review and approval.
I updated message and fix some test cases for new feature.
After CI will passed, I will squash commits into one and merge PR.

@swift-ci Please smoke test

By the way, I granted a commit access recently, but a bot doesn't respond at all.
I hope it works this time.

**Comment by owenv** [comment]
@swift-ci please smoke test

File: lib/Frontend/DiagnosticVerifier.cpp
Comment (CodaFi) [inline lib/Frontend/DiagnosticVerifier.cpp:585]: `std::tuple` is not quite the vocabulary type you want to use here. Please use a real struct with named fields so we can avoid all of that tuple unpacking noise below and get some decently self-documenting code here.

File: lib/Frontend/DiagnosticVerifier.cpp
Comment (CodaFi) [inline lib/Frontend/DiagnosticVerifier.cpp:579]: Nit: `const`

**Comment by omochi** [comment]
@CodaFi Thank you for review!
I'm grad to get your comment before merge.
I will update.

**Comment by omochi** [comment]
@shahmishal Thanks for fixing my CI trigger.

@swift-ci Please smoke test

File: lib/Frontend/DiagnosticVerifier.cpp
Comment (omochi) [reply lib/Frontend/DiagnosticVerifier.cpp]: I did it 🙂

**Comment by CodaFi** [comment]
⛵

**Comment by omochi** [comment]
Thank you all!

**Pull request merged**
